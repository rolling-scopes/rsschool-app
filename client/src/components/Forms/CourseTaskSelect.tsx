import { Select, Form } from 'antd';
import { CourseTaskDto } from 'api';
import { DeadlineIcon } from 'components/Icons/DeadlineIcon';

export enum Group {
  Default = 'default',
  Deadline = 'deadline',
  CrossCheckDeadline = 'crossCheckDeadline',
}

export type GroupType = `${Group}`;

enum Section {
  Unknown = 'Unknown',
  AllTasks = 'All tasks',
  Future = 'Future',
  Active = 'Active',
  Review = 'Review',
  Expired = 'Expired',
}

enum SortingDirection {
  Ascending = 'ascending',
  Descending = 'descending',
}

type Props = {
  data: CourseTaskDto[];
  groupBy?: GroupType;
  onChange?: (id: number) => void;
  defaultValue?: number | null;
};

export function CourseTaskSelect(props: Props) {
  const { data, groupBy = Group.Default, onChange, defaultValue, ...options } = props;
  const selectProps = onChange ? { onChange } : {};
  const selectingGroup = getSelectingGroup(data, groupBy);

  return (
    <Form.Item
      {...options}
      name="courseTaskId"
      label="Task"
      initialValue={defaultValue}
      rules={[{ required: true, message: 'Please select a task' }]}
    >
      <Select placeholder="Select task" {...selectProps} listHeight={274}>
        {selectingGroup.map(
          (section, index) =>
            section.tasks.length > 0 && (
              <Select.OptGroup key={index} label={`${section.title} (${section.tasks.length})`}>
                {section.tasks.map(task => (
                  <Select.Option key={task.id} value={task.id}>
                    <span>
                      {section.title === Section.Active && groupBy === Group.Deadline && (
                        <DeadlineIcon group={groupBy} endDate={task.studentEndDate} />
                      )}
                      {section.title === Section.Review && groupBy === Group.CrossCheckDeadline && (
                        <DeadlineIcon group={groupBy} endDate={task.crossCheckEndDate} />
                      )}{' '}
                      {task.name}
                    </span>
                  </Select.Option>
                ))}
              </Select.OptGroup>
            ),
        )}
      </Select>
    </Form.Item>
  );
}

function getSelectingGroup(data: CourseTaskDto[], groupBy: GroupType) {
  switch (groupBy) {
    case Group.Deadline:
      return [
        {
          title: Section.Active,
          tasks: sortTasks({
            tasks: getTasksBetweenStudentStartDateStudentEndDate(data),
            sortBy: 'studentEndDate',
            sortingDirection: SortingDirection.Ascending,
          }),
        },
        {
          title: Section.Future,
          tasks: sortTasks({
            tasks: getTasksBeforeStudentStartDate(data),
            sortBy: 'studentStartDate',
            sortingDirection: SortingDirection.Ascending,
          }),
        },
        {
          title: Section.Expired,
          tasks: sortTasks({
            tasks: getTasksAfterStudentEndDate(data),
            sortBy: 'studentEndDate',
            sortingDirection: SortingDirection.Descending,
          }),
        },
      ];

    case Group.CrossCheckDeadline: {
      const unknownTasks = data.filter(({ crossCheckEndDate }) => !crossCheckEndDate);
      const tasksWithoutUnknown = data.filter(({ crossCheckEndDate }) => crossCheckEndDate);

      return [
        {
          title: Section.Review,
          tasks: sortTasks({
            tasks: getTasksBetweenStudentEndDateCrossCheckEndDate(tasksWithoutUnknown),
            sortBy: 'crossCheckEndDate',
            sortingDirection: SortingDirection.Ascending,
          }),
        },
        {
          title: Section.Unknown,
          tasks: sortTasks({
            tasks: unknownTasks,
            sortBy: 'studentEndDate',
            sortingDirection: SortingDirection.Ascending,
          }),
        },
        {
          title: Section.Future,
          tasks: sortTasks({
            tasks: getTasksBeforeStudentEndDate(tasksWithoutUnknown),
            sortBy: 'studentEndDate',
            sortingDirection: SortingDirection.Ascending,
          }),
        },
        {
          title: Section.Expired,
          tasks: sortTasks({
            tasks: getTasksAfterCrossCheckEndDate(tasksWithoutUnknown),
            sortBy: 'crossCheckEndDate',
            sortingDirection: SortingDirection.Descending,
          }),
        },
      ];
    }

    case Group.Default:
    default:
      return [
        {
          title: Section.AllTasks,
          tasks: sortTasks({
            tasks: data,
            sortBy: 'studentEndDate',
            sortingDirection: SortingDirection.Descending,
          }),
        },
      ];
  }
}

type SortTasksProps = {
  tasks: CourseTaskDto[];
  sortBy: 'studentStartDate' | 'studentEndDate' | 'crossCheckEndDate';
  sortingDirection: SortingDirection;
};

function sortTasks({ tasks, sortBy, sortingDirection }: SortTasksProps) {
  return [...tasks].sort((firstTask, secondTask) => {
    const firstDate = firstTask[sortBy];
    const secondDate = secondTask[sortBy];

    if (firstDate && secondDate) {
      return sortingDirection === SortingDirection.Ascending
        ? firstDate.localeCompare(secondDate)
        : secondDate.localeCompare(firstDate);
    }

    return 1;
  });
}

function getTasksBeforeStudentStartDate(tasks: CourseTaskDto[]) {
  return tasks.filter(task => Date.now() < Date.parse(task.studentStartDate));
}

function getTasksBeforeStudentEndDate(tasks: CourseTaskDto[]) {
  return tasks.filter(task => Date.now() < Date.parse(task.studentEndDate));
}

function getTasksAfterStudentEndDate(tasks: CourseTaskDto[]) {
  return tasks.filter(task => Date.now() >= Date.parse(task.studentEndDate));
}

function getTasksAfterCrossCheckEndDate(tasks: CourseTaskDto[]) {
  return tasks.filter(task => {
    if (task.crossCheckEndDate) {
      return Date.now() >= Date.parse(task.crossCheckEndDate);
    }
  });
}

function getTasksBetweenStudentStartDateStudentEndDate(tasks: CourseTaskDto[]) {
  return tasks.filter(
    task => Date.now() >= Date.parse(task.studentStartDate) && Date.now() < Date.parse(task.studentEndDate),
  );
}

function getTasksBetweenStudentEndDateCrossCheckEndDate(tasks: CourseTaskDto[]) {
  return tasks.filter(task => {
    if (task.crossCheckEndDate) {
      return Date.now() >= Date.parse(task.studentEndDate) && Date.now() < Date.parse(task.crossCheckEndDate);
    }
  });
}
