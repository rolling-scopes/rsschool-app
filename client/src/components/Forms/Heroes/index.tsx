import { Avatar, Button, Card, Form, Grid, Input, Pagination, Row, Select, Space, Typography } from 'antd';
import { FormLayout } from 'antd/es/form/Form';
import { useCallback, useEffect, useState } from 'react';
import Masonry from 'react-masonry-css';
import css from 'styled-jsx/css';
import { IGratitudeGetRequest, IGratitudeGetResponse, HeroesFormData } from 'common/interfaces/gratitude';
import heroesBadges from 'configs/heroes-badges';
import { GratitudeService } from 'services/gratitude';
import { onlyDefined } from 'utils/onlyDefined';
import { Course } from 'services/models';
import { getFullName } from 'domain/user';

const { Text, Link, Paragraph } = Typography;
const { useBreakpoint } = Grid;

const initialPage = 1;
const initialPageSize = 20;

export const fields = {
  name: 'name',
  githubId: 'githubId',
  courseId: 'courseId',
} as const;

export const HeroesForm = ({ setLoading, courses }: { setLoading: (arg: boolean) => void; courses: Course[] }) => {
  const [heroesData, setHeroesData] = useState<IGratitudeGetResponse[]>([]);
  const [heroesCount, setHeroesCount] = useState(initialPage);
  const [currentPage, setCurrentPage] = useState(initialPage);
  const gratitudeService = new GratitudeService();
  const [form] = Form.useForm();
  const { xs } = useBreakpoint();
  const formLayout: FormLayout = xs ? 'vertical' : 'inline';
  const minWidth = xs ? undefined : 300;

  useEffect(() => {
    const getHeroes = async () => {
      setLoading(true);
      const heroes = await gratitudeService.getGratitude({ current: initialPage, pageSize: initialPageSize });
      setHeroesData(heroes.content);
      setHeroesCount(heroes.count);
      setLoading(false);
    };
    getHeroes();
  }, []);

  const makeRequest = useCallback(
    async (data: IGratitudeGetRequest) => {
      setLoading(true);
      const heroes = await gratitudeService.getGratitude(data);
      setHeroesData(heroes.content);
      setHeroesCount(heroes.count);
      setCurrentPage(initialPage);
      setLoading(false);
    },
    [heroesData],
  );

  const handleSubmit = useCallback(
    async (formData: HeroesFormData) => {
      const data = onlyDefined(formData) as Partial<HeroesFormData>;
      setCurrentPage(initialPage);
      await makeRequest(data);
    },
    [heroesData],
  );

  const onClear = useCallback(async () => {
    setLoading(true);
    const heroes = await gratitudeService.getGratitude({ current: initialPage, pageSize: initialPageSize });
    setHeroesData(heroes.content);
    setHeroesCount(heroes.count);
    setCurrentPage(initialPage);
    setLoading(false);
    form.resetFields();
  }, [heroesData]);

  const onClickPagination = useCallback(
    async (current: number, pageSize?: number) => {
      const formData = form.getFieldsValue() as HeroesFormData;
      await makeRequest({ current, pageSize: pageSize!, ...formData });
      setCurrentPage(current);
    },
    [currentPage],
  );

  return (
    <>
      <Form layout={formLayout} form={form} onFinish={handleSubmit} style={{ marginBottom: 24 }}>
        <Form.Item name={fields.name} label="Name" style={{ marginBottom: 16 }}>
          <Input />
        </Form.Item>
        <Form.Item name={fields.githubId} label="Github Username" style={{ marginBottom: 16 }}>
          <Input />
        </Form.Item>
        <Form.Item name={fields.courseId} label="Courses" style={{ minWidth, marginBottom: 16 }}>
          <Select options={courses.map(({ id, name }) => ({ value: id, label: name }))} />
        </Form.Item>
        <Space align="start" size={20}>
          <Button type="primary" htmlType="submit">
            Submit
          </Button>
          <Button type="primary" onClick={onClear}>
            Clear
          </Button>
        </Space>
      </Form>
      <Masonry
        breakpointCols={{
          default: 4,
          1100: 3,
          700: 2,
          500: 1,
        }}
        className={masonryClassName}
        columnClassName={masonryColumnClassName}
      >
        {heroesData.map(feedback => (
          <div style={{ marginBottom: gapSize, overflow: 'hidden' }} key={`card-${feedback.id}`} className="card">
            <Card style={{ position: 'relative', background: 'none' }}>
              <div
                className="badge-bg"
                style={{ backgroundImage: `url(/static/svg/badges/${heroesBadges[feedback.badgeId].url})` }}
              />
              <div className="badge-note" style={{ marginBottom: 48 }}>
                <Paragraph style={{ margin: 0 }}>
                  <Text strong>From:</Text> {getFullName(feedback.from)} (
                  <Link href={`/profile?githubId=${feedback.from.githubId}`}>@{feedback.from.githubId}</Link>)
                </Paragraph>
                <Paragraph style={{ margin: 0 }}>
                  <Text strong>To:</Text> {getFullName(feedback)} (
                  <Link href={`/profile?githubId=${feedback.githubId}`}>@{feedback.githubId}</Link>)
                </Paragraph>
              </div>
              <div className="flex-center" style={{ marginBottom: 48 }}>
                <div className="badge">
                  <Avatar
                    src={`/static/svg/badges/${heroesBadges[feedback.badgeId].url}`}
                    alt={`${feedback.badgeId} badge`}
                    size={128}
                  />
                </div>
              </div>
              <div className="badge-note">
                <Paragraph style={{ margin: 0 }}>{feedback.comment}</Paragraph>
              </div>
            </Card>
          </div>
        ))}
      </Masonry>
      {masonryStyles}
      {masonryColumnStyles}
      <Row style={{ marginTop: 16, marginBottom: 16, justifyContent: 'flex-end' }}>
        <Pagination
          current={currentPage}
          total={heroesCount}
          onChange={onClickPagination}
          defaultPageSize={initialPageSize}
        />
      </Row>
      <style jsx>{styles}</style>
    </>
  );
};

const gapSize = 16;
const { className: masonryClassName, styles: masonryStyles } = css.resolve`
  div {
    display: flex;
    margin-left: -${gapSize}px;
    width: auto;
  }
`;
const { className: masonryColumnClassName, styles: masonryColumnStyles } = css.resolve`
  div {
    padding-left: ${gapSize}px;
    background-clip: padding-box;
  }
`;

const styles = css`
  .flex-column {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
  }
  .text-wrap {
    text-align: center;
    text-overflow: ellipsis;
    width: 150px;
    overflow: hidden;
    white-space: nowrap;
  }
  .flex-center {
    display: flex;
    justify-content: center;
  }
  .badge-bg {
    position: absolute;
    background-position: center;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    z-index: -1;
    opacity: 0.1;
    transform: scale(1.2);
  }
  .badge-note {
    padding: 16px;
    background: rgba(255, 255, 255, 0.7);
    border: 2px rgba(24, 144, 255, 0.5) dashed;
  }
  .card:hover .badge {
    transform: scale(6.2);
    transition: all 1s ease;
    opacity: 0;
    z-index: -1;
  }
  .card:hover .badge-bg {
    transition: all 2s ease;
    opacity: 0.8;
  }
  .card:hover .badge-note {
    transition: all 1s ease;
    background: rgba(255, 255, 255, 1);
    border: 2px rgba(24, 144, 255, 1) dashed;
  }
`;
