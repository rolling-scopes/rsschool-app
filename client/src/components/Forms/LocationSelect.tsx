import React from 'react';
import { Select, Spin } from 'antd';
import usePlacesAutocomplete from 'use-places-autocomplete';
import { Location } from '../../../../common/models/profile';

type Props = {
  onChange: (arg: Location | null) => void;
  location: Location | null;
  style?: React.CSSProperties;
};

export function LocationSelect(props: Props) {
  const {
    value,
    suggestions: { data, loading },
    setValue,
  } = usePlacesAutocomplete({
    defaultValue: fromLocation(props.location),
    requestOptions: {
      types: ['(cities)'],
      debounce: 300,
    },
  });

  const handleInput = (value: string) => {
    setValue(value);
  };

  const handleSelect = (value: any) => {
    setValue(value, false);
    props.onChange(toLocation(value));
  };

  const handleBlur = () => {
    if (!value) {
      setValue(fromLocation(props.location), false);
    }
  };

  return (
    <Select
      filterOption={false}
      onSearch={handleInput}
      onSelect={handleSelect}
      showSearch
      onBlur={handleBlur}
      notFoundContent={loading ? <Spin size="small" /> : null}
      value={value}
      placeholder="Select users"
      options={data.map(({ description }) => ({ value: description }))}
      style={props.style}
    />
  );
}

const toLocation = (value: string): Location => {
  const parts = value.split(', ');
  return {
    cityName: parts[0],
    countryName: parts[parts.length - 1],
  };
};

const fromLocation = (value: Location | null): string => {
  if (value) {
    return `${value.cityName}, ${value.countryName}`;
  }
  return '';
};
