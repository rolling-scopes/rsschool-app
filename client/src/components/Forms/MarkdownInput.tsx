import { useEffect } from 'react';
import { Input, Form, Button, Typography } from 'antd';
import { useState } from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

type Props = {
  historicalCommentSelected: string;
};

export default function MarkdownInput({ historicalCommentSelected }: Props) {
  const [previewVisible, setPreviewVisible] = useState(false);
  const [text, setText] = useState('');

  useEffect(() => {
    if (historicalCommentSelected !== '') {
      setText(historicalCommentSelected);
    }
  }, [historicalCommentSelected]);

  const toggleView = () => {
    setPreviewVisible(!previewVisible);
  };

  const resetText = () => {
    setPreviewVisible(false);
    setText('');
  };

  const link = (
    <span style={{ marginLeft: '20px' }}>
      <a
        target="_blank"
        href="https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax"
      >
        About markdown
      </a>
    </span>
  );

  return (
    <div style={{ marginBottom: '20px' }}>
      <div style={{ display: previewVisible ? 'none' : 'block' }}>
        <Form.Item
          name="comment"
          label="Comment (markdown syntax is supported)"
          rules={[{ required: true, message: 'Please leave a detailed comment', min: 30 }]}
          onReset={resetText}
        >
          <Input.TextArea onChange={({ currentTarget: { value } }) => setText(value)} rows={5} />
        </Form.Item>
        <Button onClick={toggleView}>Preview</Button> {link}
      </div>
      {previewVisible ? (
        <div>
          <div>
            <Typography.Text>
              <ReactMarkdown rehypePlugins={[remarkGfm]}>{text}</ReactMarkdown>
            </Typography.Text>
          </div>
          <Button onClick={toggleView}>Write</Button> {link}
        </div>
      ) : (
        ''
      )}
    </div>
  );
}
