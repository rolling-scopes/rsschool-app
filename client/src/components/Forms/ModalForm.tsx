import { Form, Modal, Spin } from 'antd';
import * as React from 'react';

type Props = {
  data: any;
  title?: string;
  submit: (arg: any) => void;
  cancel: (arg: any) => void;
  onChange?: (values: any) => void;
  getInitialValues?: (arg: any) => any;
  children: any;
  loading?: boolean;
  okText?: string;
};

export function ModalForm(props: Props) {
  if (props.data == null) {
    return null;
  }
  const [form] = Form.useForm();
  const initialValues = props.getInitialValues ? props.getInitialValues?.(props.data) : props.data;
  return (
    <Modal
      style={{ top: 20 }}
      visible={true}
      title={props.title}
      okText={props.okText ?? 'Save'}
      onOk={async e => {
        e.preventDefault();
        const values = await form.validateFields().catch(() => null);
        if (values == null) {
          return;
        }
        props.submit(values);
      }}
      onCancel={e => {
        props.cancel(e);
        form.resetFields();
      }}
    >
      <Spin spinning={props.loading ?? false}>
        <Form
          onValuesChange={() => props.onChange?.(form.getFieldsValue())}
          form={form}
          initialValues={initialValues}
          layout="vertical"
        >
          {props.children}
        </Form>
      </Spin>
    </Modal>
  );
}
