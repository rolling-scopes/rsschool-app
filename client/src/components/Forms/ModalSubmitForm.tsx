import { Alert, Button, Form, Modal, Result, Spin } from 'antd';
import * as React from 'react';
type Props = {
  data: any;
  title?: string;
  submit: (arg: any) => void;
  close: (arg: any) => void;
  onChange?: (values: any) => void;
  getInitialValues?: (arg: any) => any;
  children: React.ReactNode;
  loading?: boolean;
  submitted?: boolean;
  successText?: string;
  errorText?: string;
  open?: boolean;
};

export function ModalSubmitForm({
  data,
  title,
  submit,
  close,
  onChange,
  getInitialValues,
  children,
  loading,
  submitted,
  successText,
  errorText,
  open,
}: Props) {
  if (data == null) {
    return null;
  }

  const [form] = Form.useForm();
  const initialValues = getInitialValues ? getInitialValues?.(data) : data;

  function onSubmit(): ((e: React.MouseEvent<HTMLElement, MouseEvent>) => void) | undefined {
    return async e => {
      e.preventDefault();
      const values = await form.validateFields().catch(() => null);
      if (values == null) {
        return;
      }
      submit(values);
    };
  }

  return (
    <Modal
      open={open ?? true}
      footer={submitted ? null : undefined}
      title={title}
      okText="Submit"
      onOk={onSubmit()}
      onCancel={e => {
        close(e);
        form.resetFields();
      }}
    >
      <Spin spinning={loading ?? false}>
        {errorText ? <Alert style={{ marginBottom: 16 }} message={errorText} type="error" showIcon /> : null}
        {submitted ? (
          <Result
            status="success"
            title="Success"
            subTitle={successText ?? 'Successfully submitted'}
            extra={[
              <Button style={{ minWidth: 80 }} onClick={close} type="primary" key="ok">
                Ok
              </Button>,
            ]}
          />
        ) : (
          <Form
            onValuesChange={() => onChange?.(form.getFieldsValue())}
            form={form}
            initialValues={initialValues}
            layout="vertical"
          >
            {children}
          </Form>
        )}
      </Spin>
    </Modal>
  );
}
