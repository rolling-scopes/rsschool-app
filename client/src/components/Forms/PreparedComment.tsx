import { Typography } from 'antd';
import { FC, useEffect, useState } from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

export const markdownLabel = '{markdown}\n';

type Props = {
  text: string;
};

const PreparedComment: FC<Props> = ({ text }) => {
  const [state, setState] = useState(text ?? '');
  const [isMD, setIsMD] = useState(false);

  useEffect(() => {
    if (text && text.indexOf(markdownLabel) === 0) {
      setIsMD(true);
      setState(text.slice(markdownLabel.length));
    }
  }, []);

  /**
   * In order to not affect the comments created in previosly versions of RS App, this component use a special mark for the markdown syntax.
   */
  return (
    <>
      {isMD ? (
        <Typography.Text>
          <ReactMarkdown rehypePlugins={[remarkGfm]}>{state}</ReactMarkdown>
        </Typography.Text>
      ) : (
        <Typography.Text>
          {state.split('\n').map((item, i) => (
            <div key={i}>{item}</div>
          ))}
        </Typography.Text>
      )}
    </>
  );
};

export default PreparedComment;
