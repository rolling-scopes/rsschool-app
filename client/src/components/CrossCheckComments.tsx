import { Fragment } from 'react';
import { CDN_AVATARS_URL } from 'configs/cdn';
import { ScoreIcon } from './Icons/ScoreIcon';
import { Avatar, Col, Comment, Divider, Row, Switch, Typography } from 'antd';
import { useLocalStorage } from 'react-use';
import { Feedback } from 'services/course';
import { formatDateTime } from 'services/formatter';
import { GithubUserLink } from 'components/GithubUserLink';
import PreparedComment from './Forms/PreparedComment';
import { StudentContacts } from './CrossCheck/StudentContacts';

enum LocalStorage {
  AreStudentContactsVisible = 'crossCheckAreStudentContactsVisible',
}

type Props = {
  feedback: Feedback | null;
  maxScore?: number;
};

export function CrossCheckComments({ feedback, maxScore }: Props) {
  if (!feedback || !feedback.comments || feedback.comments.length === 0) {
    return null;
  }

  const { comments } = feedback;
  const [areStudentContactsVisible = true, setAreStudentContactsHidden] = useLocalStorage<boolean>(
    LocalStorage.AreStudentContactsVisible,
  );

  const handleStudentContactsVisibilityChange = () => {
    setAreStudentContactsHidden(!areStudentContactsVisible);
  };

  return (
    <Col>
      <Row gutter={8} style={{ margin: '8px 0' }}>
        <Col>
          <Typography.Text>Student Contacts</Typography.Text>
        </Col>
        <Col>
          <Switch
            size={'small'}
            defaultChecked={areStudentContactsVisible}
            onChange={handleStudentContactsVisibilityChange}
          />
        </Col>
      </Row>

      {comments.map(({ comment, updatedDate, author, score }, i) => (
        <Fragment key={i}>
          <Row>
            <Divider style={{ margin: '8px 0' }} />
          </Row>

          <Comment
            avatar={
              <Avatar
                size={32}
                src={
                  areStudentContactsVisible && author
                    ? `${CDN_AVATARS_URL}/${author.githubId}.png?size=48`
                    : '/static/svg/badges/ThankYou.svg'
                }
              />
            }
            content={
              <>
                <Row>
                  {areStudentContactsVisible && author ? (
                    <GithubUserLink value={author.githubId} isUserIconHidden={true} />
                  ) : (
                    <Typography.Text>
                      {'Student'} {i + 1}
                      {!areStudentContactsVisible && author && ' (hidden)'}
                    </Typography.Text>
                  )}
                </Row>

                <Row>
                  <Typography.Text type="secondary" style={{ marginBottom: 8, fontSize: 12 }}>
                    {formatDateTime(updatedDate)}
                  </Typography.Text>
                </Row>

                <Row gutter={4} align="middle">
                  <Col>
                    <ScoreIcon maxScore={maxScore} score={score} />
                  </Col>
                  <Col>
                    <Typography.Text>{score}</Typography.Text>
                  </Col>
                </Row>

                <Row style={{ marginBottom: 24 }}>
                  <Typography.Text style={{ fontSize: 12, lineHeight: '12px' }} type="secondary">
                    maximum score: {maxScore ?? 'unknown'}
                  </Typography.Text>
                </Row>

                <Row>
                  <PreparedComment text={comment} />
                </Row>

                <Row>{areStudentContactsVisible && author && <StudentContacts discord={author.discord} />}</Row>
              </>
            }
          />
        </Fragment>
      ))}
      <style jsx>{`
        :global(.ant-comment-avatar) {
          position: sticky;
          top: 16px;
          align-self: start;
        }

        :global(.ant-comment-avatar img) {
          width: 100%;
          height: 100%;
        }
      `}</style>
    </Col>
  );
}
