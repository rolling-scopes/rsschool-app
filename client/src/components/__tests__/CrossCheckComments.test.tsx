import React from 'react';
import { render } from '@testing-library/react';
import { Feedback } from 'services/course';
import { markdownLabel } from 'components/Forms/PreparedComment';
import { CrossCheckComments } from '../CrossCheckComments';

describe('CrossCheckComments', () => {
  describe('Should render correctly', () => {
    const anonComment = {
      updatedDate: '2022-08-30T13:35:14.866Z',
      comment: `${markdownLabel}Awful job, you better quit programming and learn to cross stitch.`,
      score: 5,
      author: null,
    };

    const signedComment = {
      updatedDate: '2022-08-27T09:51:34.615Z',
      comment: `${markdownLabel}Great job, you better quit cross stitching and start programming.`,
      score: 100,
      author: {
        name: 'Linus Torvalds',
        githubId: 'torvalds',
        discord: {
          id: +'123456789012345678',
          username: 'LinusTorvalds',
          discriminator: 1234,
        },
      },
    };

    const emptyComments: any = [];

    const anonComments = [anonComment, anonComment, anonComment];

    const signedComments = [signedComment, signedComment, signedComment];

    const mixedComments = [anonComment, anonComment, signedComment, signedComment, anonComment];

    const feedbackWithEmptyComments: Feedback = {
      url: 'https://rolling-scopes-school.github.io/torvalds-JSFE2021Q3/async-race/',
      comments: emptyComments,
    };

    const feedbackWithAnonComments: Feedback = {
      url: 'https://rolling-scopes-school.github.io/torvalds-JSFE2021Q3/async-race/',
      comments: anonComments,
    };

    const feedbackWithSignedComments: Feedback = {
      url: 'https://rolling-scopes-school.github.io/torvalds-JSFE2021Q3/async-race/',
      comments: signedComments,
    };

    const feedbackWithMixedComments: Feedback = {
      url: 'https://rolling-scopes-school.github.io/torvalds-JSFE2021Q3/async-race/',
      comments: mixedComments,
    };

    const maxScore = 100;

    it("if there's no comments", () => {
      const output = render(<CrossCheckComments feedback={feedbackWithEmptyComments} maxScore={maxScore} />);
      expect(output.container).toMatchSnapshot();
    });
    it('if there are anonymous comments', () => {
      const output = render(<CrossCheckComments feedback={feedbackWithAnonComments} maxScore={maxScore} />);
      expect(output.container).toMatchSnapshot();
    });
    it('if there are signed comments', () => {
      const output = render(<CrossCheckComments feedback={feedbackWithSignedComments} maxScore={maxScore} />);
      expect(output.container).toMatchSnapshot();
    });
    it('if there are both types of comments', () => {
      const output = render(<CrossCheckComments feedback={feedbackWithMixedComments} maxScore={maxScore} />);
      expect(output.container).toMatchSnapshot();
    });
  });
});
