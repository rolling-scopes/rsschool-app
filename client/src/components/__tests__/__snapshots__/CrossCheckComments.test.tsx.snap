// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrossCheckComments Should render correctly if there are anonymous comments 1`] = `
<Col>
  <Row
    key="0"
  >
    <Divider />
    <Comment
      author="Student 1"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="1"
  >
    <Divider />
    <Comment
      author="Student 2"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="2"
  >
    <Divider />
    <Comment
      author="Student 3"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
</Col>
`;

exports[`CrossCheckComments Should render correctly if there are both types of comments 1`] = `
<Col>
  <Row
    key="0"
  >
    <Divider />
    <Comment
      author="Student 1"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="1"
  >
    <Divider />
    <Comment
      author="Student 2"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="2"
  >
    <Divider />
    <Comment
      author={
        <GithubUserLink
          value="torvalds"
        />
      }
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            100
          </Paragraph>
          <PreparedComment
            text="Great job, you better quit cross stitching and start programming."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="3"
  >
    <Divider />
    <Comment
      author={
        <GithubUserLink
          value="torvalds"
        />
      }
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            100
          </Paragraph>
          <PreparedComment
            text="Great job, you better quit cross stitching and start programming."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="4"
  >
    <Divider />
    <Comment
      author="Student 5"
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            5
          </Paragraph>
          <PreparedComment
            text="Awful job, you better quit programming and learn to cross stitch."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
</Col>
`;

exports[`CrossCheckComments Should render correctly if there are signed comments 1`] = `
<Col>
  <Row
    key="0"
  >
    <Divider />
    <Comment
      author={
        <GithubUserLink
          value="torvalds"
        />
      }
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            100
          </Paragraph>
          <PreparedComment
            text="Great job, you better quit cross stitching and start programming."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="1"
  >
    <Divider />
    <Comment
      author={
        <GithubUserLink
          value="torvalds"
        />
      }
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            100
          </Paragraph>
          <PreparedComment
            text="Great job, you better quit cross stitching and start programming."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
  <Row
    key="2"
  >
    <Divider />
    <Comment
      author={
        <GithubUserLink
          value="torvalds"
        />
      }
      avatar={<ForwardRef(UserOutlined) />}
      content={
        <React.Fragment>
          <Paragraph
            strong={true}
          >
            Score: 
            100
          </Paragraph>
          <PreparedComment
            text="Great job, you better quit cross stitching and start programming."
          />
        </React.Fragment>
      }
      style={
        Object {
          "margin": 16,
        }
      }
    />
  </Row>
</Col>
`;

exports[`CrossCheckComments Should render correctly if there's no comments 1`] = `""`;
